src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs

```diff
@@ -0,0 +1,262 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Options;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Options.Providers;
+using Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+using SVsServiceProvider = Microsoft.VisualStudio.Shell.SVsServiceProvider;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Options
+{
+    /// <summary>
+    /// An <see cref="IOptionPersister"/> that syncs core language settings against the settings that exist for all languages
+    /// in Visual Studio and whose backing store is provided by the shell. This includes things like default tab size, tabs vs. spaces, etc.
+    /// </summary>
+    [Export(typeof(IOptionPersister))]
+    internal sealed class LanguageSettingsPersister : ForegroundThreadAffinitizedObject, IVsTextManagerEvents4, IOptionPersister
+    {
+        private readonly IVsTextManager4 _textManager;
+        private readonly IGlobalOptionService _optionService;
+
+        private readonly IComEventSink _textManagerEvents2Sink;
+
+        private readonly IBidirectionalMap<string, Guid> _languageMap;
+
+        /// <remarks>We make sure this code is from the UI by asking for all serializers on the UI thread in <see cref="HACK_AbstractCreateServicesOnUiThread"/>.</remarks>
+        [ImportingConstructor]
+        public LanguageSettingsPersister([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, IGlobalOptionService optionService)
+            : base(assertIsForeground: true)
+        {
+            _textManager = (IVsTextManager4)serviceProvider.GetService(typeof(SVsTextManager));
+            _optionService = optionService;
+
+            // TODO: make this configurable
+            _languageMap = BidirectionalMap<string, Guid>.Empty.Add(LanguageNames.CSharp, Guids.CSharpLanguageServiceId)
+                                                               .Add(LanguageNames.VisualBasic, Guids.VisualBasicLanguageServiceId);
+
+            foreach (var languageGuid in _languageMap.Values)
+            {
+                var languagePreferences = new LANGPREFERENCES3[1];
+                languagePreferences[0].guidLang = languageGuid;
+                Marshal.ThrowExceptionForHR(_textManager.GetUserPreferences4(null, languagePreferences, null));
```

Comment:
Added named parameters for the nulls. You can fetch several different values with this function (view and color settings), the nulls say we don't care about those.

I believe this should be fine, but good call on the perf question; testing now.